packages/amplify-graphql-function-transformer/src/graphql-function-transformer.ts

```diff
@@ -0,0 +1,174 @@
+import { DirectiveWrapper, MappingTemplate, TransformerPluginBase } from '@aws-amplify/graphql-transformer-core';
+import { TransformerContextProvider, TransformerSchemaVisitStepContextProvider } from '@aws-amplify/graphql-transformer-interfaces';
+import * as lambda from '@aws-cdk/aws-lambda';
+import * as cdk from '@aws-cdk/core';
+import { obj, str, ref, printBlock, compoundExpression, qref, raw, iff } from 'graphql-mapping-template';
+import { FunctionResourceIDs, ResolverResourceIDs, ResourceConstants } from 'graphql-transformer-common';
+import { DirectiveNode, ObjectTypeDefinitionNode, InterfaceTypeDefinitionNode, FieldDefinitionNode } from 'graphql';
+
+type FunctionDirectiveConfiguration = {
+  name: string;
+  region: string | undefined;
+  resolverTypeName: string;
+  resolverFieldName: string;
+};
+
+const FUNCTION_DIRECTIVE_STACK = 'FunctionDirectiveStack';
+const directiveDefinition = /* GraphQL */ `
+  directive @function(name: String!, region: String) repeatable on FIELD_DEFINITION
+`;
+
+export class FunctionTransformer extends TransformerPluginBase {
+  private resolverGroups: Map<FieldDefinitionNode, FunctionDirectiveConfiguration[]> = new Map();
+
+  constructor() {
+    super('amplify-function-transformer', directiveDefinition);
+  }
+
+  field = (
+    parent: ObjectTypeDefinitionNode | InterfaceTypeDefinitionNode,
+    definition: FieldDefinitionNode,
+    directive: DirectiveNode,
+    acc: TransformerSchemaVisitStepContextProvider,
+  ): void => {
+    const directiveWrapped = new DirectiveWrapper(directive);
+    const args = directiveWrapped.getArguments({
+      resolverTypeName: parent.name.value,
+      resolverFieldName: definition.name.value,
+    } as FunctionDirectiveConfiguration);
+    let resolver = this.resolverGroups.get(definition);
+
+    if (resolver === undefined) {
+      resolver = [];
+      this.resolverGroups.set(definition, resolver);
+    }
+
+    resolver.push(args);
+  };
+
+  generateResolvers = (context: TransformerContextProvider): void => {
+    if (this.resolverGroups.size === 0) {
+      return;
+    }
+
+    const stack: cdk.Stack = context.stackManager.createStack(FUNCTION_DIRECTIVE_STACK);
```

Comment:
I'd like to avoid character counting if possible. It would also be nice if the related data source, app sync function, and resolvers could be kept together in the same file to avoid awkward dependencies between files.

What do you think about splitting this up at the data source level? That would ensure that there are no dependencies across files. It would still be theoretically possible to hit a limit, but I don't think that can ever be avoided if an app is sufficiently large.

Reply:
> I'd like to avoid character counting if possible. It would also be nice if the related data source, app sync function, and resolvers could be kept together in the same file to avoid awkward dependencies between files.

I don't char counting either. One approach we could take is to say how many function we put in a single stack, and if we exceed the number of functions then we create another stack for function. I am not sure what would the ideal number of function in a stack. We could potentially change the number of function in future if we want without causing any issue as the functions are just data sources and by themself they can be moved around without causing any data loss.

> What do you think about splitting this up at the data source level? That would ensure that there are no dependencies across files. It would still be theoretically possible to hit a limit, but I don't think that can ever be avoided if an app is sufficiently large.

I am not sure if I understand this correctly, but with `@function` the DataSource is a lambda function right. Are you suggesting all the `@function` that use the same lambda function go in a single stack?

Reply:
> Are you suggesting all the @function that use the same lambda function go in a single stack?

Yes. So, in the following example, there would be a single data source, a single appsync function configuration, and two pipeline resolvers. I was proposing trying to keep all of that together in a single file.

```graphql
type Query {
    echo(msg: String): String @function(name: "echofunction-\${env}")
    magic(msg: String): String @function(name: "echofunction-\${env}")
}
```

> One approach we could take is to say how many function we put in a single stack, and if we exceed the number of functions then we create another stack for function.

That would work too, and should be fairly straightforward.